usr/src/uts/common/fs/zfs/dmu_objset.c

```diff
@@ -632,7 +632,6 @@ dmu_objset_refresh_ownership(objset_t *os, void *tag)
 	dsl_pool_config_enter(dp, FTAG);
 	dmu_objset_disown(os, tag);
 	VERIFY0(dsl_dataset_own(dp, name, tag, &newds));
-	VERIFY3P(newds, ==, os->os_dsl_dataset);
```

Comment:
the verify is correct and should not be removed.  otherwise os_dsl_dataset could be some random pointer and not the actual dataset that we own.  It seems like we are relying on dsl_dataset_hold_obj()'s call to dmu_buf_get_user() to still have the dataset cached.  Maybe @scsiguy's changes to the dbuf user stuff changed these semantics?  The fix may be as simple as setting os_dsl_dataset to newds, but we would have to see if anything else assumes that os_dsl_dataset does not change.